### PR TITLE
refactor reader for readability and flexible sets of features

### PR DESCRIPTION
**Pull request by sunya-ch**
This PR refactor the source reader.go with the following items.
- separate pod_energy.go and node_energy.go to handle PodEnergy and CurrNodeEnergy struct separately from the reader.
- use UInt64Stat (Curr,Aggr) struct instead of two static variables to handle these two values systematically. (including overflow checking)
- use map of metric(feature) name for not only cgroup but also hardware counters and kubelet metrics.
- use model/estimator.go to compute pod power

There are two kinds of powers computed.
- **Ratio power:** This is a package power computed by ratio of usage metric defined in config module to the RAPL power.
- **Dynamic power:** This is a dynamic part of package power predicted by customized or trained model if available using only usage metric .

remarks:
- Instead of compute core power from pkg power, I think we should report as it is. If RAPL energy for core component is not available, we should not report it.
- This should work even no estimator sidecar container. (no dynamic power reporter)

Signed-off-by: Sunyanan Choochotkaew <sunyanan.choochotkaew1@ibm.com>

### CODE REVIEW

**Comment by rootfs** [comment]
thank you @sunya-ch 

can you post the the prometheus query result sample?

File: pkg/collector/collector.go [hunk 1]
```diff
@@ -286,131 +230,60 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
...
+				v.PodName, v.Namespace,
+			)
+			ch <- edesc
+		}
+		// dyn curr pod energy
+		edesc := prometheus.MustNewConstMetric(
+			podDesc.CurrEnergy[DYN_ENERGY_LABEL],
 			prometheus.GaugeValue,
-			float64(v.CurrEnergyInOther),
+			float64(v.DynEnergy.Curr)/1000.0, //J
```
Comment (rootfs) [inline pkg/collector/collector.go:265]: If switching to J from mJ, please also update the dashboard unit

**Comment by sunya-ch** [comment]
> thank you @sunya-ch
> 
> can you post the node and pod prometheus query result sample?

Theses are pod's result. (total and curr in RAPL pkg)
Screenshot 2022-08-23 at 21 50 15

Screenshot 2022-08-23 at 22 17 07

And, I just add node energy report with amended commit. (at the beginning, it has only node stat)
Screenshot 2022-08-23 at 22 12 53

File: pkg/collector/collector.go (hunk 1 above)
Comment (sunya-ch) [reply pkg/collector/collector.go:297]: I'm sorry but could you clarify where do you mean the dashboard unit?

Comment (rootfs) [reply pkg/collector/collector.go:297]: the dashboard uses mW https://github.com/sustainable-computing-io/kepler/blob/main/grafana-dashboards/Kepler-Exporter-Dashboard-openshift.json#L133

Comment (sunya-ch) [reply pkg/collector/collector.go:297]: I see. Do you think should we export energy in J or mJ? 
If you agree to change to J, I will update the grafana. But, if you think it is better to keep the decimal, I will modify the collector sourcecode.

Comment (rootfs) [reply pkg/collector/collector.go:297]: let's keep mJ then, thanks

Comment (sunya-ch) [reply pkg/collector/collector.go:297]: Thanks. I see that originally you report node power as J. Should I keep it like that? 
So, the units will be
- node power -> J
- package power -> mJ
- pod power -> mJ 

do I understand it correctly?

Comment (rootfs) [reply pkg/collector/collector.go]: that sounds right

**Comment by rootfs** [comment]
thank you @sunya-ch

File: pkg/collector/reader.go [hunk 2]
```diff
@@ -52,459 +52,331 @@ type CgroupTime struct {
...
-				coreDelta := float64(energyCore-lastEnergyCore) / samples
-				dramDelta := float64(energyDram-lastEnergyDram) / samples
-				if coreDelta == 0 && dramDelta == 0 {
-					log.Printf("power reading not changed, retry\n")
-					samples += 1
-					lock.Unlock()
-					continue
-				}
-
-				if coreDelta > maxEnergyDelta || dramDelta > maxEnergyDelta {
```
Comment (rootfs) [inline pkg/collector/reader.go:204]: is the sanity checking still applicable?

**Pull request merged**